### PR TITLE
[WIP] ipvs ebpf support non masqurade cidrs

### PR DESCRIPTION
**Pull request by kenieevan**

### CODE REVIEW

**Comment by tencent-adm** [comment]
[CLA assistant check](https://cla.opensource.tencent.com/Tencent/TencentOS-kernel?pullRequest=18) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.tencent.com/Tencent/TencentOS-kernel?pullRequest=18) before we can accept your contribution.<br/><hr/>**jianmingfan** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.tencent.com/check/Tencent/TencentOS-kernel?pullRequest=18) it.</sub>

File: net/netfilter/ipvs/ip_vs_bpf_proc.c [hunk 1]
```diff
@@ -854,11 +1050,17 @@ int ip_vs_svc_proc_init(struct netns_ipvs *ipvs)
...
 	remove_proc_entry("ip_vs_svc_ip_attr", ipvs->net->proc_net);
 	remove_proc_entry("ip_vs_bpf_id", ipvs->net->proc_net);
 	remove_proc_entry("ip_vs_devip", ipvs->net->proc_net);
 	remove_proc_entry("ip_vs_port_range", ipvs->net->proc_net);
 	remove_proc_entry("ip_vs_bpf_stat", ipvs->net->proc_net);
 	remove_proc_entry("ip_vs_bpf_fix", ipvs->net->proc_net);
+	remove_proc_entry("ip_vs_non_masq_cidrs", ipvs->net->proc_net);
+	rcu_read_lock();
+	old = rcu_dereference(non_masq_cidrs);
+	rcu_read_unlock();
```
Comment (honkiko) [inline net/netfilter/ipvs/ip_vs_bpf_proc.c:1063]: 这里要加  synchronize_rcu();

File: net/netfilter/ipvs/ip_vs_conn.c [hunk 2]
```diff
@@ -1069,6 +1060,34 @@ static void nf_conntrack_single_unlock(struct bpf_lb_conn_key *key,
 	spin_unlock_bh(&bpf_conntrack_locks[hash]);
 }
 
+
+/* ip in host endian */
+bool ip_in_vpc(u32 ip)
```
Comment (honkiko) [inline net/netfilter/ipvs/ip_vs_conn.c:1065]: 建议换个名字。non-masq-cidr是任意配置的，不一定跟vpc有关系

File: net/netfilter/ipvs/ip_vs_bpf_proc.c [hunk 3]
```diff
@@ -815,8 +815,199 @@ static const struct file_operations ip_vs_bpf_stat_fops = {
...
+
+	memset(buf, 0, sizeof(buf));
+	rcu_read_lock();
+	c = rcu_dereference(non_masq_cidrs);
+	if (!c) {
+		rcu_read_unlock();
+		return -EFAULT;
+	}
+	for (i = 0; i < c->len; i++) {
+		/* snprintf need tail to save null bytes */
```
Comment (gxm-newton) [inline net/netfilter/ipvs/ip_vs_bpf_proc.c:861]: c->len会有多长？ 如果比较多，需要再循环里加cond_resched

File: net/netfilter/ipvs/ip_vs_bpf_proc.c [hunk 4]
```diff
@@ -815,8 +815,199 @@ static const struct file_operations ip_vs_bpf_stat_fops = {
...
+	if (count <= 0 || count > MAXCIDRNUM * CIDRLEN) {
+		pr_err("%s %d count %ld\n", __func__, __LINE__, count);
+		return -EINVAL;
+	}
+	buf = kzalloc(MAXCIDRNUM * CIDRLEN, GFP_KERNEL);
+	if (!buf)
+		return -ENOMEM;
+
+	new = kzalloc(sizeof(struct cidrs), GFP_KERNEL);
+	if (!new)
```
Comment (gxm-newton) [inline net/netfilter/ipvs/ip_vs_bpf_proc.c]: 这里直接return， buf就泄露了，建议加个err_out标签，集中处理出错回滚

File: net/netfilter/ipvs/ip_vs_bpf_proc.c (hunk 3 above)
Comment (liuxu623) [reply net/netfilter/ipvs/ip_vs_bpf_proc.c:861]: #define MAXCIDRNUM 64
这个最长应该不超过64

File: net/netfilter/ipvs/ip_vs_bpf_proc.c (hunk 4 above)
Comment (kenieevan) [reply net/netfilter/ipvs/ip_vs_bpf_proc.c]: ok.fixed in new pull request

**Pull request closed without merging**